File: edukate-messaging/src/main/java/io/github/sanyavertolet/edukate/messaging/RabbitAutoConfiguration.java
package io.github.sanyavertolet.edukate.messaging;

import org.springframework.amqp.core.*;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.boot.autoconfigure.AutoConfiguration;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
import org.springframework.context.annotation.Bean;

@AutoConfiguration
@ConditionalOnClass(AmqpAdmin.class)
public class RabbitAutoConfiguration {

    @Bean
    @ConditionalOnMissingBean
    public TopicExchange edukateExchange() {
        return new TopicExchange(RabbitTopology.EXCHANGE, true, false);
    }

    @Bean
    @ConditionalOnMissingBean
    public Jackson2JsonMessageConverter jackson2JsonMessageConverter() {
        return new Jackson2JsonMessageConverter();
    }
}

File: edukate-messaging/src/main/kotlin/io/github/sanyavertolet/edukate/messaging/RabbitAutoConfiguration.kt
package io.github.sanyavertolet.edukate.messaging

import org.springframework.amqp.core.AmqpAdmin
import org.springframework.amqp.core.TopicExchange
import org.springframework.amqp.support.converter.JacksonJsonMessageConverter
import org.springframework.boot.autoconfigure.AutoConfiguration
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean
import org.springframework.context.annotation.Bean

@AutoConfiguration
@ConditionalOnClass(AmqpAdmin::class)
class RabbitAutoConfiguration {
    @Bean @ConditionalOnMissingBean fun edukateExchange() = TopicExchange(RabbitTopology.EXCHANGE, true, false)

    @Bean @ConditionalOnMissingBean fun jacksonJsonMessageConverter() = JacksonJsonMessageConverter()
}
